ambry-api/src/main/java/com.github.ambry/frontend/ContainerMetrics.java

```diff
@@ -0,0 +1,121 @@
+   * For example:
+   * {@code com.github.ambry.frontend.ContainerMetrics.account-a___container-b___GetBlobSuccessCount}
+   * @param accountName the account name to use for naming metrics.
+   * @param containerName the container name to use for naming metrics.
+   * @param operationType the operation type to use for naming metrics.
+   * @param metricRegistry the {@link MetricRegistry}.
+   */
+  ContainerMetrics(String accountName, String containerName, String operationType, MetricRegistry metricRegistry) {
+    String metricPrefix = accountName + SEPARATOR + containerName + SEPARATOR + operationType;
+    roundTripTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "RoundTripTimeInMs"));
+
+    successCount = metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "SuccessCount"));
+    redirectionCount =
+        metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "RedirectionCount"));
+    clientErrorCount =
+        metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "ClientErrorCount"));
+    serverErrorCount =
+        metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "ServerErrorCount"));
+
+    badRequestCount =
+        metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "BadRequestCount"));
+    unauthorizedCount =
+        metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "UnauthorizedCount"));
+    forbiddenCount =
+        metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "ForbiddenCount"));
+    notFoundCount = metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "NotFoundCount"));
+    goneCount = metricRegistry.counter(MetricRegistry.name(ContainerMetrics.class, metricPrefix + "GoneCount"));
+  }
+
+  /**
+   * Emit metrics for an operation on this container.
+   * @param roundTripTimeInMs the time it took to receive a request and send a response.
+   * @param responseStatus the {@link ResponseStatus} sent in the response.
+   */
+  public void recordMetrics(long roundTripTimeInMs, ResponseStatus responseStatus) {
+    this.roundTripTimeInMs.update(roundTripTimeInMs);
+
+    if (responseStatus.isSuccess()) {
+      successCount.inc();
+    } else if (responseStatus.isRedirection()) {
+      redirectionCount.inc();
+    } else if (responseStatus.isClientError()) {
+      clientErrorCount.inc();
+      switch (responseStatus) {
+        case BadRequest:
+          badRequestCount.inc();
+          break;
+        case Unauthorized:
+          unauthorizedCount.inc();
+          break;
+        case Forbidden:
+          forbiddenCount.inc();
+          break;
+        case NotFound:
+          notFoundCount.inc();
+          break;
+        case Gone:
+          goneCount.inc();
+          break;
+      }
```

Comment:
Default in case new statuses get added?

Reply:
This switch isn't inclusive of all error codes, only the ones that customers expressed interest in monitoring. So, I'm not sure if there would be anything that would go in a default block. The count of all the other 4xx error code can be computed with `clientErrorCount - badRequestCount - unauthorizedCount - forbiddenCount - notFoundCount - goneCount`

Reply:
Or do you think it's better if we get rid of the overall `clientErrorCount` and make a count for the default case, `otherClientErrorCount`

Reply:
It seems fine as it is.  If nothing needs to be done in default block and it's not an error case, then it can be skipped.